### PR TITLE
feat(util): Create generic parser tools

### PR DESCRIPTION
**Pull request by zharinov**
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
## Roadmap

### Tokenizer

- Space parsing options:
  - [ ] Group spaces and tabs together with newlines separated
  - [ ] Group all spaces into single token
  - [ ] Separate token for each character: do we need this?
- Symbols (identifiers and keywords):
  - [ ] Configure by raw regex
  - [ ] Alphanumeric defaults + customize allowed characters, e.g. `$` for JS or `-`, `>`, `<`, `!` for Clojure
- String literals:
  - [ ] Resolve opening char sequence ordering, e.g. `"` and `"""`
    - [x] Implement generic sorting function for such kind of tokenizer rules
  - [ ] Character escaping ([comprehensive list](https://en.wikipedia.org/wiki/String_literal#Escape_sequences)).
  - Templates
    - [ ] Resolve opening char sequence (sigil) ordering, e.g. `$foo` vs `${foo}`
    - [ ] Explicit termination, e.g. `${` and `}`
    - [ ] Implicit termination, e.g. `"$groupId:$artifactId:$version"`
      - [ ] Configure token subset, e.g. allow dots for `"org.apache:lucene-core:$Versions.lucene"`
  - [Here documents](https://en.wikipedia.org/wiki/Here_document): 
    - Mostly not useful for Renovate purposes, but can significantly poison parsing results
    - Bad news: implementation is tricky
      - [ ] Configure sigil
      - [ ] Implement specific transformer:
        1. Detect special token that matches `[sigil]+[heredoc-content]+[heredoc-end]+[rest-of-file]`
        2. Save underlying tokenizer state
        3. Output `[heredoc]` token with text value of `[sigil]+[heredoc-content]+[heredoc-end]`
        4. Restore underlying  tokenizer state starting with `[heredoc-end]`
    - Good news: it seems to be low priority
      - Languages that support it:
        - **Ruby**
        - PHP
        - Perl
        - Unix shells
        - ...

### Parser

- Guiding principles:
  - Extract structural features for interesting parts, combining simple tokens into some higher-level elements
  - Ignore other tokens, leaving them as-is
  - Use token sequence transformers for particular language aspects
    - Stateful
    - Composable
    - Examples:
      - Heredocs string tokens (see Tokenizer section above)
      - `[quotes]+[chars]+[quotes]` to `[String]`
      - `[quotes]+[chars]+[sigil]+[symbol]+[chars]+[quotes]` to `[StringTemplate]`
      - `[newline]+[space]+[space]` to `[newline]+[Indent]` or `[newline]+[Indent]+[Indent]` (depending on previous indents)
      - Produce `[Instruction]` sequences based on newlines or semicolons
      - Produce nested `[Block]` objects based on indents or braces
  - Match on parsed structure using selector-based approach (like CSS or XPath):
    - `[Block]` nesting, e.g. `deps` section:
      - `str('deps').child(str('*', 'lval').op('=').str('rval')).many()`
    - `[Instruction]` internal structure, e.g. variable assignment or `registryUrl` definition:
      - `sym('*', 'lval').op(':=').str('rval')`
      - `sym('url').wrap(str('registryUrl'), opt = true, brackets = 'circle')`
    - Some similar DSLs:
      - [**parsimmon**](https://github.com/jneen/parsimmon/blob/master/API.md) parser combinators
      - [**@thi.ng/zipper**](https://github.com/thi-ng/umbrella/tree/develop/packages/zipper) library for functional zippers for TypeScript

File: lib/util/parser/tokenizer-config/index.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import { fallback as fallbackRule } from 'moo';
+import type { Rules as MooRules } from 'moo';
+import { TokenType as t } from '../token';
+import { commentRule } from './comment';
+import type { CommentOption, TokenizerOptions } from './types';
+import { ensureArray, ruleName, sortOptions } from './utils';
+
+interface MooStates {
+  [x: string]: MooRules;
+}
```
[the remark this answers is not in the document]
Comment (zharinov) [reply lib/util/parser/tokenizer-config/index.ts:10]: It's just the copy-paste from 'moo' anonymous input type for `moo.states()`. Yes, can be a `Record`

**Comment by rarkins** [comment]
@zharinov what were your plans with this?

**Comment by zharinov** [comment]
I was thinking of separating this from main Renovate repository, as it should have a lot of unit tests

**Comment by rarkins** [comment]
OK, let's close for now

**Pull request closed without merging**